Hardware accelerated implementation of Quaternion multiply

TJHeuvel opened the pull request:
I saw on the Discord that Quaternion multiply wasnt yet vectorised, this PR adds that. The non accelerated path is the exact same as before, and the accelerated path returns the same results.

TJHeuvel committed:
Vector based implementation of Quaternion multiply

I saw on the Discord that Quaternion multiply wasnt yet vectorised, this PR adds that. The non accelerated path is the exact same as before, and the accelerated path appears to return the same results.

TJHeuvel commented:
@dotnet-policy-service agree

TJHeuvel committed:
Convert tabs to spaces

AustinWise commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:
```diff
@@ -188,31 +188,49 @@ public static Quaternion Identity
+                float q2w = value1.W;
+
+                Vector3 q1V = new Vector3(value2.X, value2.Y, value2.Z);
+                Vector3 q2V = new Vector3(value1.X, value2.Y, value2.Z);
```
```suggestion
                Vector3 q2V = new Vector3(value1.X, value1.Y, value1.Z);
```

I don't know anything about quaternions, but is this a copy-paste error?

TJHeuvel replied:
Thank you, you're completely right!

TJHeuvel committed:
Update src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs

Copy  paste error!

Co-authored-by: Austin Wise <AustinWise@gmail.com>

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:
```diff
@@ -188,31 +188,49 @@ public static Quaternion Identity
-            ans.W = q1w * q2w - dot;
-
-            return ans;
+            if (Vector.IsHardwareAccelerated)
+            {
+                // Concatenate rotation is actually q2 * q1 instead of q1 * q2.
+                // So that's why value2 goes q1 and value1 goes q2.
+                float q1w = value2.W;
+                float q2w = value1.W;
+
+                Vector3 q1V = new Vector3(value2.X, value2.Y, value2.Z);
+                Vector3 q2V = new Vector3(value1.X, value1.Y, value1.Z);
+
+                return new Quaternion(
+                    q1w * q2V + q2w * q1V + Vector3.Cross(q1V, q2V),
+                    q1w * q2w - Vector3.Dot(q1V, q2V)
+                );
+            }
```
I would expect something like the following to perform better:
```csharp
if (Vector128.IsHardwareAccelerated)
{
    var left = value1.AsVector128();
    var right = value2.AsVector128();

    var result = right * left.GetElement(3);
    result += (Vector128.Shuffle(right, Vector128.Create(3, 2, 1, 0)) * left.GetElement(0)) * Vector128.Create(+1.0f, -1.0f, +1.0f, -1.0f);
    result += (Vector128.Shuffle(right, Vector128.Create(2, 3, 0, 1)) * left.GetElement(1)) * Vector128.Create(+1.0f, +1.0f, -1.0f, -1.0f);
    result += (Vector128.Shuffle(right, Vector128.Create(1, 0, 3, 2)) * left.GetElement(2)) * Vector128.Create(-1.0f, +1.0f, +1.0f, -1.0f);
    return result.AsQuaternion();
}
```

TJHeuvel replied:
Thanks, that is indeed quite a lot faster, and the results still looked correct.

```
| Method                 | Mean      | Error     | StdDev    |
|----------------------- |----------:|----------:|----------:|
| MulOperator            | 4.5998 ns | 0.0843 ns | 0.0788 ns |
| VectorMul              | 2.9167 ns | 0.0145 ns | 0.0128 ns |
| IntrinsicMulGetElement | 0.8482 ns | 0.0113 ns | 0.0106 ns |
```

I've modified it to use the GetElementUnsafe method, i saw other Numerics types use this too.

TJHeuvel committed:
Even faster version

Tanner provided an even quicker version. I've changed it to use GetElementUnsafe as i saw other API's doing the same, and we can reasonably argue these fields to exist.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:
```diff
@@ -188,31 +188,45 @@ public static Quaternion Identity
+                result += (Vector128.Shuffle(right, Vector128.Create(3, 2, 1, 0)) * left.GetElementUnsafe(0)) * Vector128.Create(+1.0f, -1.0f, +1.0f, -1.0f);
+                result += (Vector128.Shuffle(right, Vector128.Create(2, 3, 0, 1)) * left.GetElementUnsafe(1)) * Vector128.Create(+1.0f, +1.0f, -1.0f, -1.0f);
+                result += (Vector128.Shuffle(right, Vector128.Create(1, 0, 3, 2)) * left.GetElementUnsafe(2)) * Vector128.Create(-1.0f, +1.0f, +1.0f, -1.0f);
+                return result.AsQuaternion();
```
Looks like this will need to be
```suggestion
                return Unsafe.BitCast<Vector128<float>, Quaternion>(result);
```

I had *thought* I added `AsVector128` and `AsQuaternion` at the same time, but apparently I only added one. I'm going to open an API proposal to get them public and to cover the missing inverse conversions at the same time.

TJHeuvel replied:
My apologies, i just wanted to do a quick improvement and dont have a great development environment to catch this. I also indeed saw that the `AsQuaternion` existed as internal and assumed it was good. 

Should this PR use the bitcast, or would it be better to wait for the API proposal?

tannergooding replied:
I'd just use BitCast for now to unblock the PR and then we can go and fix it up later after the API becomes usable.

TJHeuvel committed:
Update src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs

Co-authored-by: Tanner Gooding <tagoo@outlook.com>

TJHeuvel commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:
```diff
@@ -188,31 +188,45 @@ public static Quaternion Identity
+            {
+                // Concatenate rotation is actually q2 * q1 instead of q1 * q2.
+                // So that's why value2 goes q1 and value1 goes q2.
+                float q1x = value2.X;
+                float q1y = value2.Y;
+                float q1z = value2.Z;
+                float q1w = value2.W;
+
+                float q2x = value1.X;
+                float q2y = value1.Y;
+                float q2z = value1.Z;
+                float q2w = value1.W;
```
I dont think i understand, the original version also swaps these around. value2 is loaded into q1, there is a comment explaining why. 

The non-hardware accelerated version has not changed as far as i can tell.

Rob-Hague replied:
@TJHeuvel I think you might have inadvertently replaced the non-hardware accelerated implementation with that from `Concatenate`. The values are not swapped in the `operator *` implementation

TJHeuvel replied:
Yes absolutely, my bad

TJHeuvel committed:
Fix incorrect non-hardware accelerated version

tannergooding commented:
Failures are unrelated issues that are already tracked.

cincuranet commented:
Looks like we have some small number of regressions for this (rest is not interesting): https://github.com/dotnet/perf-autofiling-issues/issues/27922

cc @TJHeuvel @tannergooding
